### PR TITLE
discard-duplicates: option to keep first

### PR DESCRIPTION
**Pull request by christophehurpeau**
example:

```
.ResetButton { padding: 0 }
.SpecificButton { padding: 20px }
.ResetButton { padding: 0 }
```

without `keepFirst`:

```
.SpecificButton { padding: 20px }
.ResetButton { padding: 0 }
```

with `keepFirst: true`: 

```
.ResetButton { padding: 0 }
.SpecificButton { padding: 20px }
```

### CODE REVIEW

**Comment by evilebottnawi** [comment]
Why we need this option?

**Comment by ben-eb** [comment]
```html
<div class="SpecificButton ResetButton">
  Button
</div>
```

```css
div {
  background: red
}

.ResetButton { padding: 0 }
.SpecificButton { padding: 20px }
.ResetButton { padding: 0 }
```

In this case if the second rule is discarded the button has `20px` padding, whereas the default behaviour preserves the `0` padding.

**Comment by evilebottnawi** [comment]
@ben-eb do you think we need this option?

**Comment by ben-eb** [comment]
@evilebottnawi I'd argue no, unless there's a good reason that I've missed.

**Comment by evilebottnawi** [comment]
@ben-eb thanks for feedback, let's wait answer

**Comment by christophehurpeau** [comment]
With css modules, we have 2 files that imports the ResetButton with composes. I don't see another solution for this problem ?

**Comment by evilebottnawi** [comment]
@christophehurpeau change order of classes?

**Comment by christophehurpeau** [comment]
@evilebottnawi I don't understand your answer ?

**Comment by anikethsaha** [comment]
I think this will create an inconsistency with the styling order.
-1 for this,

> @evilebottnawi I don't understand your answer?

@christophehurpeau you need to change the order of the class name where you are using these class in order to get the required styling

**Comment by christophehurpeau** [comment]
@anikethsaha I don't understand, the order of classes is right (.ResetButton then .SpecificButton) the problem is that another css modules uses the reset css file, reimporting .ResetButton after .SpecificButton.
We cannot not import the reset file in the second file, because it needs it too.

The css is generated like this:
```
.ResetButton { padding: 0 }
.SpecificButton1 { padding: 20px }
.ResetButton { padding: 0 }
.SpecificButton2 { padding: 20px }
```

In my opinion, in that case, only the first `.ResetButton` should be kept, not the last.

**Comment by anikethsaha** [comment]
```css
.ResetButton {
  padding: 0;
}
.SpecificButton {
  padding: 20px;
}
```

now if you have HTML element class using these

```html
<button class="ResetButton SpecificButton ">its 20px</button>
<button class="SpecificButton  ResetButton ">its 0px</button>
```

I am not getting your doubt. Is it like you are getting a process CSS and you cant change the order ?

**Comment by christophehurpeau** [comment]
@anikethsaha in my understanding, changing the order in the dom doen't affect the applied css, only the css order matters

**Comment by anikethsaha** [comment]
yeah my bad 😁

**Comment by evilebottnawi** [comment]
> changing the order in the html doen't affect the applied css

I disagree, clarify what do you mean

**Comment by anikethsaha** [comment]
```html
<style>
  .a {
    color: red;
  }
  .b {
    color: blue;
  }
</style>

<p class="b a">hello</p> <!-- blue -->
<p class="a b">hello</p> <!-- blue -->
```

cc @evilebottnawi

**Comment by evilebottnawi** [comment]
@anikethsaha yep, you are right, but i can't understand why we need this option

**Comment by anikethsaha** [comment]
> @anikethsaha yep, you are right, but I can't understand why we need this option

yeah same doubt for me too

> I think its more like a use-case dependent issue.

It should be changed with the input, not by cssnano !
order stuff is user-dependent, cause cssnano can't predict how it's being used in the HTML elements

@christophehurpeau do you have any strong reason to support this ?

**Comment by christophehurpeau** [comment]
The reason is explained here: https://github.com/cssnano/cssnano/pull/809#issuecomment-553802579
I have two components that both uses the same css file with css modules `composes`
The result is something like:
```

**Comment by anikethsaha** [comment]
I didnt use css module as of now of I guess thats what keeping me in doubt with your use-case.
I hope others might review this properly

Thanks 👍

**Comment by evilebottnawi** [comment]
@chriseppstein please create reproducible test repo with css modules configuration

**Comment by tallytarik** [comment]
How can I help keep this moving? Really need this feature at the moment!

**Comment by anikethsaha** [comment]
As far as the above discussion is concerned, the use case is not clear. 

We would appreciate it if you can create an issue with repro may be using a repo or sandbox and also a use case might move the discussion forward.

Thanks

**Comment by sigveio** [comment]
@tallytarik Due to no follow up in the past year and the amount of time that has passed since this PR was submitted, I'm closing it as stale. You are welcome to [open an issue](https://github.com/cssnano/cssnano/issues/new/choose) to discuss further if you believe there is a valid use case for the feature you suggested. Thanks =)
